File: 2001-3000/2353-design-a-food-rating-system/2353-design-a-food-rating-system.cpp
class FoodRatings {
private:
    // Custom comparator for the set to sort by rating (descending) 
    // and then by food name (ascending/lexicographically smaller).
    struct Compare {
        bool operator()(const pair<int, string>& a, const pair<int, string>& b) const {
            if (a.first != b.first) {
                return a.first > b.first; // Higher rating comes first
            }
            return a.second < b.second;   // Alphabetically smaller name comes first
        }
    };

    // Maps food name -> pair(rating, cuisine)
    unordered_map<string, pair<int, string>> foodMap;
    
    // Maps cuisine -> sorted set of pair(rating, food name)
    unordered_map<string, set<pair<int, string>, Compare>> cuisineMap;

public:
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        int n = foods.size();
        for (int i = 0; i < n; ++i) {
            foodMap[foods[i]] = {ratings[i], cuisines[i]};
            cuisineMap[cuisines[i]].insert({ratings[i], foods[i]});
        }
    }
    
    void changeRating(string food, int newRating) {
        // Get the cuisine type for this food
        string cuisine = foodMap[food].second;
        
        // Update the master rating in our map
        foodMap[food].first = newRating;
        
        // Insert the newly updated pair into the cuisine's sorted set
        cuisineMap[cuisine].insert({newRating, food});
    }
    
    string highestRated(string cuisine) {
        // Look at the top element of the set for this cuisine
        auto& highestSet = cuisineMap[cuisine];
        
        // Clean up any outdated "ghost" ratings at the top of the set
        while (!highestSet.empty()) {
            auto topElement = *(highestSet.begin());
            string foodName = topElement.second;
            int setRating = topElement.first;
            
            // If the rating in the set matches the actual current rating, it's valid!
            if (foodMap[foodName].first == setRating) {
                return foodName;
            }
            
            // Otherwise, it's an old rating. Delete it and keep checking.
            highestSet.erase(highestSet.begin());
        }
        
        return "";
    }
};

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings* obj = new FoodRatings(foods, cuisines, ratings);
 * obj->changeRating(food,newRating);
 * string param_2 = obj->highestRated(cuisine);
 */

File: 2001-3000/2353-design-a-food-rating-system/2353-design-a-food-rating-system.java
class FoodRatings {

    public record FoodRating(String food, String cuisine, int rating) {}

    Map<String, FoodRating> foodRatings = new HashMap<>(); // food -> FoodRating
    Map<String, PriorityQueue<FoodRating>> foodsByCuisine = new HashMap<>();

    public FoodRatings(String[] foods, String[] cuisines, int[] ratings) {
        for(int i=0; i<foods.length; i++) {
            add(new FoodRating(foods[i], cuisines[i], ratings[i]));
        }
    }

    void add(FoodRating f) {
        foodRatings.put(f.food(), f);
        foodsByCuisine.computeIfAbsent(f.cuisine(), k -> new PriorityQueue<>((a,b) -> {
            if (a.rating() != b.rating()) return b.rating() - a.rating(); // highest rating first
            return a.food().compareTo(b.food()); // If there is a tie, return the item with the lexicographically smaller name.
        })).add(f);
    }
    
    public void changeRating(String food, int newRating) {
        FoodRating f = foodRatings.get(food);
        if (f != null) {
            add(new FoodRating(food, f.cuisine(), newRating));
        }
    }
    
    public String highestRated(String cuisine) {
        PriorityQueue<FoodRating> pq = foodsByCuisine.get(cuisine);
        if (pq != null) {
            while(!pq.isEmpty()) {
                FoodRating f = pq.peek();
                if (foodRatings.get(f.food()) == f) return f.food();
                pq.poll();
            }
            // cleanup empty q -- this should not happen
            foodsByCuisine.remove(cuisine);
        }
        return "";
    }
}

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings obj = new FoodRatings(foods, cuisines, ratings);
 * obj.changeRating(food,newRating);
 * String param_2 = obj.highestRated(cuisine);
 */